README
Bongo is a flexible and usable media player for GNU Emacs.  If you
store your music collection locally and use structured file names,
then Bongo is a great way for you to play music from within Emacs.

Bongo can be installed through MELPA using `M-x package-install'.

Getting Started
===============

Though Bongo has certain preferences about how you name the files in
your music collection (see below), you should be able to get started
right away by runing `M-x bongo RET' and following the instructions.

Before getting your hands dirty, you might want to skim through the
customization group for Bongo (`M-x customize-group RET bongo RET').
It might be fun to browse through all of it, but at the moment the
most important option is `bongo-enabled-backends', which determines
what external programs Bongo will use to play media.

Bongo will try to find a reasonable set of default backends by looking
for the programs `vlc', `mpg123', `ogg123', `speexdec', `timidity' and
`mikmod'.  (If you install one of these programs now, you will have to
restart Emacs for Bongo to pick it up automatically.)

If you want to use another program for media playback, you can easily
define your own backend.  Basic play/pause functionality works with
any backend, whereas other features like seeking require custom code.
(The default pause functionality is implemented using `SIGSTOP'.)

Your Music Collection
=====================

For the best Bongo user experience, your music collection should have
hierarchical file names.  That is, they should be named according to
one of the following templates:

   Artist - 2014 - Album - 01 - Title.mp3
   Artist - 2014 - Album - Title.mp3
   Artist - Album - 01 - Title.mp3
   Artist - Album - Title.mp3
   Artist - Title.mp3
   Title.mp3

So any reasonable combination of artist, year, album, index and title
in the right order (less specific to more specific) should be okay.
Embedding the metadata in the file name is simple and robust and also
speeds up the process of fetching it.

You don't have to use ` - ' as the separator, but you can't use `/'
and have your files in nested directories.  The information actually
needs to be stored in the basename of each file.

Many people use tags (e.g., ID3v2 tags or Ogg tags) for their metadata
and don't care about file names.  Luckily, Bongo comes with a tool
called `tree-from-tags.rb', which can be used to create a hierarchy of
nicely-named symlinks into your tagged media collection.

Basic Commands
==============

To start Bongo, run `M-x bongo RET'.  To insert a file or a whole
directory of files, type `i'.  Then hit `RET' to play one of them.
To see a list of available commands, run `M-x describe-mode RET'.
Here are some common ones for your convenience:

  `i'         Insert a track (or a directory tree full of tracks).
  `RET'       Start playing the track at point (and then continue).
  `SPC'       Pause playback (if supported by the backend).
  `C-c C-s'   Stop playback (or start, if already stopped).
  `C-c C-n'   Start playing the next track.
  `C-c C-r'   Start playing a random track.
  `s'         Seek backward or forward (if supported by the backend).
  `l'         Move point to the current track and recenter.
  `q'         Quit Bongo (bury Bongo buffers and delete windows).

  `c'         Copy the track or section under point.
  `k'         Kill the track or section under point.
  `C-w'       Kill all tracks and sections in the region.
  `y'         Reinsert the last copied or killed stuff.
  `I u'       Insert a URI (can be used to play podcasts and radio).
  `I TAB'     List other kinds of tracks that can be inserted.

  `2 C-c C-n' Skip the next track and start playing the one after that.
  `3 C-c C-s' Stop playback after the next three tracks finish playing.
  `3 RET'     Start playing the track under point, and continue playing
              tracks, but stop after playing the third track.
  `C-u C-c C-s'
              Stop playback after each track finishes playing (this is
              good when playing movies, for example).
  `C-u C-u C-c C-s'
              Stop playback after the track at point finishes playing.
  `C-u C-c C-n'
              Play the next track after each track finishes playing
              (this undoes the effect of `C-u C-c C-s').
  `C-u C-c C-r'
              Play tracks in random order.

Here are some commands that are nice if you use library buffers:

  `h'         Switch from playlist to library (or vice versa).
  `C-u h'     Switch from playlist to library (or vice versa), leaving
              the original buffer visible in another window.
  `e'         Append the track under point or the region (if active)
              to the end of the playlist buffer.
  `3 e'       Append the track at point and the two after that to the
              end of the playlist buffer.
  `E'         Insert the track under point or the region (if active)
              into the playlist buffer, directly below the track
              that's currently playing.
  `C-x C-s'   Save the current buffer (library buffers are stored
              using the `.bongo-library' file extension).

Finally, since you've read this far, here are a few bonus commands:

  `d'         Open a Dired buffer containing the track under point.
  `r'         Rename the file corresponding to the track under point.
  `v'         Change audio volume (requires the `volume-el' package).


NEWS
Bongo NEWS file listing user-visible changes.

May 14, 2014

 * `bongo-insert-album-covers' now defaults to nil.

May 12, 2014

 * Bongo can now be installed from MELPA using `M-x package-install'.

 * Bongo no longer supports Emacs 21.

 * All deprecated variables and functions have been purged.

 * New simple backend `afplay' adds support for OS X.

July 13, 2009

 * Bongo now forces VLC to exit after finishing a track.

 * `C-u RET' on a section header now starts playing
   the section.  To set the backend, use `C-u C-u RET'.

May 21, 2007

 * New key bindings `S-<up>' and `S-<down>' (each with `M-'
   and `C-M-' variants) change the volume.

 * New key bindings `S-<right>' and `S-<left>' (with `M-'
   and `C-M-' variants) fast-forward and rewind.

 * The `f' and `b' keys have been switched with `F' and `B'.
   Now use `f' or `b' to seek just one second.  (This lets
   you seek N seconds using `<N> f' or `<N> b'.)

 * The <TAB> (`bongo-toggle-collapsed') command has been
   moved to `$' to make room for a future <TAB> command.

 * New command `H' (`bongo-list-buffers') brings up a list
   of all Bongo buffers to ease switching between them.

 * Giving a prefix argument to `h' (`bongo-switch-buffers')
   now makes Bongo prompt for the buffer to switch to.

 * Now <mouse-3> in Bongo buffers always brings up a context
   menu (it used to do it on object lines only).

 * Files with `.bongo' extensions are now treated as either
   playlists or libraries according to the value of the user
   option `bongo-prefer-library-buffers'.

 * Switching to a Bongo mode from a non-fundamental,
   non-Bongo mode now requires confirmation.

 * New minor mode `bongo-dired-library-mode' lets you use
   any Dired buffer as a Bongo library.  The mode is enabled
   automatically by the `d' (`bongo-dired-line') command.

May 15, 2007

 * The `tree-from-tabs.rb' script has a new command line
   option, `--hardlinks', which tells it to create hardlinks
   instead of symbolic links.

May 11, 2007

 * There is now a volume control mode line button.  Click it
   to open the volume control interface (as `v' would), or
   use your mouse wheel on it to adjust the volume directly.

   You should update the `volume' library to get the most
   out of this feature.  Mouse support was recently added.

May 6, 2007

 * The keys `f' and `b' are now bound to seeking commands,
   to go with the newly-introduced `F', `B', `M-F', etc.

   The old `f' command (`bongo-flush-playlist') has been
   moved to `~' (played tracks are "backups", in a sense).

 * The command `C-c TAB' now sits on `C-c C-e' instead,
   because of the symmetry with `C-c C-a'.

April 28, 2007

 * The keys `F', `B', `M-F', `M-B', `C-M-F' and `C-M-B' are
   now bound to immediate seeking commands, which provide
   quicker ways of seeking (you no longer have to use `s').

   The old `F' command has been moved to `R' (for "reset").

 * Inline playback progress is now disabled by default,
   because it slowed down seeking.  To turn it back on,
   customize `bongo-display-inline-playback-progress'.

 * Left-clicking the arrow buttons in the mode line now
   fast-forwards or rewinds instead of switching tracks ---
   to switch tracks, use the right mouse button (`mouse-3').

 * The arrow buttons will now only appear during playback
   (leaving just the play button when playback is stopped).

   To pick your own set of mode line buttons, customize the
   variable `bongo-mode-line-indicator-format'.

April 27, 2007

 * The keys in seek mode have changed a little.  It used to
   be that `f' and `b' each sought one second, while `M-f'
   and `M-b' sought ten seconds.  Here are the new keys:

    - One second: `C-f', `C-b', `F', `B'
    - Three seconds: `f', `b', `<left>', `<right>'
    - Ten seconds: `M-f', `M-b', `M-<left>', `M-<right>'
    - One minute: `C-M-f', `C-M-b'

   Some other keys in this keymap changed as well, but I
   don't think any of it is worth mentioning.

April 26, 2007

 * Playback progress is now shown directly in the playlist
   buffer using the `bongo-elapsed-track-part' face (which
   by default looks like the `bongo-played-track' face).

 * Track lengths are now displayed in playlist buffers.
   You may want to customize `bongo-track-length-column' or
   (to disable this feature) `bongo-display-track-lengths'.

   [This feature is still rather experimental and probably
   does not work the way we will ultimately want it to.]

April 11, 2007

 * There is a new binding for playing stuff: `g'.  It plays
   the next N tracks, given a prefix argument N, or the
   tracks in the region, if the region is active, or the
   marked tracks, if any, or the track at point otherwise.

   The `RET' command is unchanged; it still plays the track
   under point, or the next N tracks with a prefix argument.
   (Note that `RET' never plays, e.g., all marked tracks.)

April 10, 2007

 * You may now bring up a context menu of useful commands by
   right-clicking a track line or section header.  Try it!

April 9, 2007

 * Bongo now displays any metadata provided by streaming
   media servers (such as what's currently being played).
   This is enabled by default but only available for VLC.

April 5, 2007

 * Bongo Sprinkle mode now sets `bongo-mark-played-tracks',
   and the default value is now `nil' (it used to be `t').
   Enabling Sprinkle mode sets it to `t' buffer-locally.

 * Stopping a track being played using `C-c C-s' will now
   not mark it as played; but `C-c C-n' and `C-c C-p' will.

April 4, 2007

 * The new Bongo Sprinkle mode is an alternative to random
   playback, based on the idea of automatically _enqueuing_
   random tracks instead of playing tracks in random order.
   To try it, run `M-x bongo-sprinkle-mode' in a playlist.

 * The various commands beginning with `i' are no longer
   bound to any keys.  Instead, there is a general `i'
   command for inserting files and directories, and another

bongo-lastfm.el
;;; bongo-lastfm.el --- submit information to Last.fm from Bongo
;; Copyright (C) 2006  Daniel Brockman

;; Author: Daniel Brockman <daniel@brockman.se>
;; URL: http://www.brockman.se/software/bongo/bongo-lastfm.el
;; Created: May 24, 2006
;; Updated: September 19, 2006

;; This file is part of Bongo.

;; This file is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License as
;; published by the Free Software Foundation; either version 2 of
;; the License, or (at your option) any later version.

;; This file is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty
;; of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
;; See the GNU General Public License for more details.

;; You should have received a copy of the GNU General Public
;; License along with GNU Emacs; if not, write to the Free
;; Software Foundation, 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:

;; This file requires the `lastfm-submit' library,
;; available at the following URL:

;; <http://www.brockman.se/software/lastfm-submit.el>

;;; Code:

(require 'lastfm-submit)

(defun bongo-lastfm-cancel-timer (player)
  (when (bongo-player-get player 'lastfm-timer)
    (cancel-timer (bongo-player-get player 'lastfm-timer))
    (bongo-player-put player 'lastfm-timer nil)))

(defun bongo-lastfm-tick (player)
  ;; The Audioscrobbler website says that each song should
  ;; be submitted ``when it is 50% or 240 seconds complete,
  ;; whichever comes first.''
  (when (or (>= (bongo-player-elapsed-time player)
                (/ (bongo-player-total-time player) 2.0))
            (>= (bongo-player-elapsed-time player) 240))
    (when (or (null (bongo-player-buffer player))
              (with-current-buffer (bongo-player-buffer player)
                bongo-lastfm-mode))
      (bongo-lastfm-submit-player player))
    (bongo-lastfm-cancel-timer player)))

(defun bongo-lastfm-player-started-1 (player)
  (when (and (bongo-player-elapsed-time player)
             (bongo-player-total-time player)
             ;; ``Songs with a duration of less than 30
             ;; seconds should not be submitted,'' says the
             ;; Audioscrobbler website.
             (>= (bongo-player-total-time player) 30))
    (let ((timer (run-with-timer 1 1 'bongo-lastfm-tick player)))
      (bongo-player-put player 'lastfm-timer timer))))

(defun bongo-lastfm-player-started (player)
  (let ((timer (run-with-timer 5 nil 'bongo-lastfm-player-started-1 player)))
    (bongo-player-put player 'lastfm-timer timer)))

(defun bongo-lastfm-player-disrupted (player)
  (bongo-lastfm-cancel-timer player))

(define-minor-mode bongo-lastfm-mode
  "Toggle Bongo Last.fm mode in the current buffer.
Interactively, with no prefix argument, toggle the mode.
With universal prefix ARG, turn mode on.
With zero or negative ARG, turn mode off.
In Bongo Last.fm mode, song information is automatically
sumbitted to Last.fm (using `lastfm-submit')."
  :lighter " Last.fm"
  (when (not (bongo-playlist-buffer-p))
    (setq bongo-lastfm-mode nil)
    (error "Bongo Last.fm mode can only be enabled in Bongo playlists"))
  (if bongo-lastfm-mode
      (progn
        (add-hook 'bongo-player-started-functions
                  'bongo-lastfm-player-started nil 'local)
        (add-hook 'bongo-player-sought-functions
                  'bongo-lastfm-player-disrupted nil 'local)
        (add-hook 'bongo-player-stopped-functions
                  'bongo-lastfm-player-disrupted nil 'local))
    (remove-hook 'bongo-player-started-functions
                 'bongo-lastfm-player-started 'local)
    (remove-hook 'bongo-player-sought-functions
                 'bongo-lastfm-player-disrupted 'local)
    (remove-hook 'bongo-player-stopped-functions
                 'bongo-lastfm-player-disrupted 'local)))

(defun bongo-lastfm-submit (infoset length)
  "Submit song information to Last.fm using `lastfm-submit'."
  (when (null length)
    (error "Can't submit track to Last.fm: missing song length"))
  (lastfm-submit (bongo-infoset-artist-name infoset)
                 (bongo-infoset-track-title infoset)
                 (number-to-string (round length))
                 (bongo-infoset-album-title infoset))
  (message "Submitted to Last.fm: %s"
           (bongo-format-infoset infoset)))

(defun bongo-lastfm-submit-player (player)
  "Submit PLAYER's song information to Last.fm.
See `bongo-lastfm-submit'."
  (bongo-lastfm-submit (bongo-infoset-from-file-name
                        (bongo-player-file-name player))
                       (bongo-player-total-time player)))

(defun bongo-lastfm-submit-current ()
  "Sumbit the currently playing track to Last.fm."
  (interactive)
  (with-bongo-playlist-buffer
    (if (bongo-playing-p)
        (bongo-lastfm-submit-player bongo-player)
      (error "No active track"))))

;;; Local Variables:
;;; coding: utf-8
;;; time-stamp-format: "%:b %:d, %:y"
;;; time-stamp-start: ";; Updated: "
;;; time-stamp-end: "$"
;;; End:

(provide 'bongo-lastfm)
;;; bongo-lastfm.el ends here.


bongo-emacs21.el
;;; bongo-emacs21.el --- compatibility functions for Emacs 21
;; Copyright (C) 2006, 2007  Daniel Brockman
;; Copyright (C) 1985, 1986, 1992, 1994, 1995, 1996, 1997, 1999, 2000,
;;   2001, 2002, 2003, 2004, 2005, 2006  Free Software Foundation, Inc.

;; Author: Daniel Brockman <daniel@brockman.se>
;; Created: December 27, 2006

;; This file is part of Bongo.

;; Bongo is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License as
;; published by the Free Software Foundation; either version 2 of
;; the License, or (at your option) any later version.

;; Bongo is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty
;; of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
;; See the GNU General Public License for more details.

;; You should have received a copy of the GNU General Public
;; License along with Bongo (see the file `COPYING');
;; if not, write to the Free Software Foundation, 51 Franklin
;; Street, Fifth Floor, Boston, MA 02110-1301, USA.

;;; Commentary:

;; This file contains functions that are necessary for Bongo
;; to work with Emacs 21.  Most definitions were simply
;; copied from Emacs 22 and then backported to Emacs 21.

;;; Code:

(eval-when-compile
  (require 'cl))

(defun bongo-face-foreground (face &optional frame inherit)
  "Call `face-foreground' with FACE and FRAME.
INHERIT is ignored, since it is not supported by Emacs 21."
  (face-foreground face frame))

(defun bongo-face-background (face &optional frame inherit)
  "Call `face-background' with FACE and FRAME.
INHERIT is ignored, since it is not supported by Emacs 21."
  (face-background face frame))

(defalias 'bongo-read-directory-name
  'read-file-name)

;;; The following function was copied from `subr.el'
;;; and modified to work with Emacs 21.

(defun bongo-run-mode-hooks (&rest hooks)
  "Run HOOKS and then `after-change-major-mode-hook'.
Major mode functions should use this."
  (apply 'run-hooks hooks)
  (run-hooks 'after-change-major-mode-hook))

;;; The following function was copied from `subr.el'.

(defun bongo-read-number (prompt &optional default)
  (let ((n nil))
    (when default
      (setq prompt
	    (if (string-match "\\(\\):[ \t]*\\'" prompt)
		(replace-match (format " (default %s)" default) t t prompt 1)
	      (replace-regexp-in-string "[ \t]*\\'"
					(format " (default %s) " default)
					prompt t t))))
    (while
	(progn
	  (let ((str (read-from-minibuffer prompt nil nil nil nil
					   (and default
						(number-to-string default)))))
	    (setq n (cond
		     ((zerop (length str)) default)
		     ((stringp str) (read str)))))
	  (unless (numberp n)
	    (message "Please enter a number.")
	    (sit-for 1)
	    t)))
    n))

;;; The following were copied from `image.el.'

(defconst bongo-image-type-file-name-regexps
  '(("\\.png\\'" . png)
    ("\\.gif\\'" . gif)
    ("\\.jpe?g\\'" . jpeg)
    ("\\.bmp\\'" . bmp)
    ("\\.xpm\\'" . xpm)
    ("\\.pbm\\'" . pbm)
    ("\\.xbm\\'" . xbm)
    ("\\.ps\\'" . postscript)
    ("\\.tiff?\\'" . tiff))
  "Alist of (REGEXP . IMAGE-TYPE) pairs used to identify image files.
When the name of an image file match REGEXP, it is assumed to
be of image type IMAGE-TYPE.")

(defun bongo-image-type-from-file-name (file)
  "Determine the type of image file FILE from its name.
Value is a symbol specifying the image type, or nil if type cannot
be determined."
  (let ((types bongo-image-type-file-name-regexps)
	type)
    (while types
      (if (string-match (car (car types)) file)
	  (setq type (cdr (car types))
		types nil)
	(setq types (cdr types))))
    type))

;;; The following macros were copied from `byte-run.el'.

(defmacro bongo-define-obsolete-function-alias
  (obsolete-name current-name &optional when docstring)
  "Make OBSOLETE-NAME an obsolete function alias for CURRENT-NAME.
This just calls `defalias' and `make-obsolete'.
DOCSTRING is ignored, because `defalias' does not accept the
corresponding argument in Emacs 21."
  (declare (doc-string 4))
  `(progn
     (defalias ,obsolete-name ,current-name)
     (make-obsolete ,obsolete-name ,current-name ,when)))

(defmacro bongo-define-obsolete-variable-alias
  (obsolete-name current-name &optional when docstring)
  "Make OBSOLETE-NAME an obsolete variable.
This just calls `make-obsolete-variable'.  It doesn't actually create a
variable alias, because there is no such thing in Emacs 21."
  (declare (doc-string 4))
  `(progn

;;; Not available in Emacs 21.
;;;      (defvaralias ,obsolete-name ,current-name ,docstring)

     (make-obsolete-variable ,obsolete-name ,current-name ,when)))


;;;; Fallback implementations of `process-{get,put}'.

(defvar bongo-process-alist nil)

(defun bongo-process-plist (process)
  (bongo-alist-get bongo-process-alist process))

(defun bongo-process-set-plist (process plist)
  (bongo-alist-put 'bongo-process-alist process plist))

(defun bongo-process-get (process property)
  "Return the value of PROPERTY for PROCESS."
  (plist-get (bongo-process-plist process) property))

(defun bongo-process-put (process property value)
  "Change the value of PROPERTY for PROCESS to VALUE."
  (bongo-process-set-plist
   process (plist-put (bongo-process-plist process)
                      property value)))


;;;; Custom

(require 'custom)

;;; The following functions were copied from `custom.el'
;;; and modified to work with Emacs 21.

(defun bongo-custom-set-minor-mode (variable value)
  ":set function for minor mode variables.
Normally, this sets the default value of VARIABLE to nil if VALUE
is nil and to t otherwise, but if `custom-local-buffer' is non-nil,
this sets the local binding in that buffer instead."
  (if custom-local-buffer
      (with-current-buffer custom-local-buffer
	(funcall variable (if value 1 0)))
    (funcall variable (if value 1 0))))

(defun bongo-customize-mark-as-set (symbol)
  "Mark current value of SYMBOL as being set from customize.

If the default value of SYMBOL is different from the saved value if any,
or else if it is different from the standard value, set the
`customized-value' property to a list whose car evaluates to the
default value.  Otherwise, set it to nil.

Return non-nil iff the `customized-value' property actually changed."

;;; Not available in Emacs 21.
;;;   (custom-load-symbol symbol)

  (require 'cus-edit)
  (let* ((get (or (get symbol 'custom-get) 'default-value))
         (value (funcall get symbol))
         (customized (get symbol 'customized-value))
         (old (or (get symbol 'saved-value) (get symbol 'standard-value))))
    ;; Mark default value as set iff different from old value.
    (if (not (and old
                  (equal value (condition-case nil
                                   (eval (car old))
                                 (error nil)))))
        (progn (put symbol 'customized-value (list (custom-quote value)))

;;; Not available in Emacs 21.
;;;                (custom-push-theme 'theme-value symbol 'user 'set
;;;                                   (custom-quote value))

               )
      (put symbol 'customized-value nil))
    ;; Changed?
    (not (equal customized (get symbol 'customized-value)))))

(defun bongo-custom-reevaluate-setting (symbol)
  "Reset the value of SYMBOL by re-evaluating its saved or standard value.
Use the :set function to do so.  This is useful for customizable options
that are defined before their standard value can really be computed."
  (funcall (or (get symbol 'custom-set) 'set-default)
           symbol
           (eval (car (or (get symbol 'saved-value)
                          (get symbol 'standard-value))))))


;;;; Define global minor mode

;;; The following function was copied from `easy-mmode.el'.

(defmacro bongo-define-global-minor-mode
  (global-mode mode turn-on &rest keys)
  "Make GLOBAL-MODE out of the buffer-local minor MODE.
TURN-ON is a function that will be called with no args in every buffer
  and that should try to turn MODE on if applicable for that buffer.
KEYS is a list of CL-style keyword arguments.  As the minor mode
  defined by this function is always global, any :global keyword is
  ignored.  Other keywords have the same meaning as in `define-minor-mode',
  which see.  In particular, :group specifies the custom group.
  The most useful keywords are those that are passed on to the
  `defcustom'.  It normally makes no sense to pass the :lighter
  or :keymap keywords to `define-global-minor-mode', since these
  are usually passed to the buffer-local version of the minor mode.

If MODE's set-up depends on the major mode in effect when it was
enabled, then disabling and reenabling MODE should make MODE work
correctly with the current major mode.  This is important to
prevent problems with derived modes, that is, major modes that
call another major mode in their body."

  (let* ((global-mode-name (symbol-name global-mode))
	 (pretty-name (easy-mmode-pretty-mode-name mode))
	 (pretty-global-name (easy-mmode-pretty-mode-name global-mode))
	 (group nil)
	 (extra-keywords nil)
	 (MODE-buffers (intern (concat global-mode-name "-buffers")))
	 (MODE-enable-in-buffers
	  (intern (concat global-mode-name "-enable-in-buffers")))
	 (MODE-check-buffers
	  (intern (concat global-mode-name "-check-buffers")))
	 (MODE-cmhh (intern (concat global-mode-name "-cmhh")))
	 (MODE-major-mode (intern (concat (symbol-name mode) "-major-mode")))
	 keyw)

    ;; Check keys.
    (while (keywordp (setq keyw (car keys)))
      (setq keys (cdr keys))
      (case keyw
	(:group (setq group (nconc group (list :group (pop keys)))))
	(:global (setq keys (cdr keys)))
	(t (push keyw extra-keywords) (push (pop keys) extra-keywords))))

    (unless group
      ;; We might as well provide a best-guess default group.
      (setq group
	    `(:group ',(intern (replace-regexp-in-string
				"-mode\\'" "" (symbol-name mode))))))

    `(progn
       (defvar ,MODE-major-mode nil)
       (make-variable-buffer-local ',MODE-major-mode)
       ;; The actual global minor-mode
       (define-minor-mode ,global-mode
	 ,(format "Toggle %s in every buffer.
With prefix ARG, turn %s on if and only if ARG is positive.
%s is actually not turned on in every buffer but only in those
in which `%s' turns it on."
		  pretty-name pretty-global-name pretty-name turn-on)
	 :global t ,@group ,@(nreverse extra-keywords)

	 ;; Setup hook to handle future mode changes and new buffers.
	 (if ,global-mode
	     (progn
	       (add-hook 'after-change-major-mode-hook
			 ',MODE-enable-in-buffers)
	       (add-hook 'find-file-hook ',MODE-check-buffers)
	       (add-hook 'change-major-mode-hook ',MODE-cmhh))
	   (remove-hook 'after-change-major-mode-hook ',MODE-enable-in-buffers)
	   (remove-hook 'find-file-hook ',MODE-check-buffers)
	   (remove-hook 'change-major-mode-hook ',MODE-cmhh))

	 ;; Go through existing buffers.
	 (dolist (buf (buffer-list))
	   (with-current-buffer buf
	     (if ,global-mode (,turn-on) (when ,mode (,mode -1))))))

       ;; Autoloading define-global-minor-mode autoloads everything
       ;; up-to-here.
       :autoload-end

       ;; List of buffers left to process.
       (defvar ,MODE-buffers nil)

       ;; The function that calls TURN-ON in each buffer.
       (defun ,MODE-enable-in-buffers ()
	 (dolist (buf ,MODE-buffers)
	   (when (buffer-live-p buf)
	     (with-current-buffer buf
	       (if ,mode
		   (unless (eq ,MODE-major-mode major-mode)
		     (,mode -1)
		     (,turn-on)
		     (setq ,MODE-major-mode major-mode))
		 (,turn-on)
		 (setq ,MODE-major-mode major-mode))))))
       (put ',MODE-enable-in-buffers 'definition-name ',global-mode)

       (defun ,MODE-check-buffers ()
	 (,MODE-enable-in-buffers)
	 (setq ,MODE-buffers nil)
	 (remove-hook 'post-command-hook ',MODE-check-buffers))
       (put ',MODE-check-buffers 'definition-name ',global-mode)

       ;; The function that catches kill-all-local-variables.
       (defun ,MODE-cmhh ()
	 (add-to-list ',MODE-buffers (current-buffer))
	 (add-hook 'post-command-hook ',MODE-check-buffers))
       (put ',MODE-cmhh 'definition-name ',global-mode))))

;;; Local Variables:
;;; coding: utf-8
;;; End:

(provide 'bongo-emacs21)
;;; bongo-emacs21.el ends here.


tree-from-tags.rb
#!/usr/bin/env ruby
## tree-from-tags.rb --- create file hierarchies from media tags
# Copyright (C) 2006, 2007  Daniel Brockman

# Author: Daniel Brockman <daniel@brockman.se>
# Created: April 24, 2006

# This file is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of
# the License, or (at your option) any later version.
#
# This file is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty
# of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public
# License along with GNU Emacs; if not, write to the Free
# Software Foundation, 51 Franklin Street, Fifth Floor,
# Boston, MA 02110-1301, USA.
#
# To run this program, you need Ruby-taglib, available at
# <http://www.hakubi.us/ruby-taglib/>, and Mahoro, available at
# <http://mahoro.rubyforge.org/>.

require "fileutils"
require "find"
require "taglib"

if ARGV.empty? or ["-?", "-h", "-help", "--help"].include? ARGV.first
  puts "Usage: #$0 [--hardlinks] DIRECTORIES..."
  puts "
This program recursively scans DIRECTORIES for media files in formats
that support embedded file tags, such as Ogg and MP3.

For each file with sufficient embedded information, it creates a symlink
in the current directory, or in a subdirectory of the current directory,
pointing to the original file.

If given the `--hardlinks' option, it creates hardlinks instead.

The symlinks or hardlinks created by this program follow a certain
naming scheme that is understood by Bongo, the Emacs media player."
  exit
end

if ["--hard", "--hardlink", "--hardlinks"].include? ARGV.first
  ARGV.shift
  $hardlinks = true
else
  $hardlinks = false
end

class NotEnoughData < RuntimeError ; end

class String
  def blank? ; !self[/\S/] end
  def trim ; sub(/^\s+|\s+$/, "") end
end

def escape_component (component)
  case component
  when "." then "Dot"
  when ".." then "Double Dot"
  else component.gsub(" - ", " -- ").gsub("/", "\\")
  end
end

def join_components (*components)
  components.compact * " - "
end

def parse_data (data)
  an = data[:artist_name]
  ay = data[:album_year]
  at = data[:album_title]
  ti = data[:track_index]
  tt = data[:track_title]

  ay = nil if ay == "0"

  case
  when an && ay && at && ti && tt
    components = [[an], [ay, at], [ti, tt]]
  when an && ay && at && tt
    components = [[an], [ay, at], [tt]]
  when an && at && ti && tt
    components = [[an], [at], [ti, tt]]
  when an && at && tt
    components = [[an], [at], [tt]]
  when an && tt
    components = [[an], [tt]]
  when tt
    components = [[tt]]
  else raise NotEnoughData
  end

  return components.map { |x| x.map { |x| escape_component(x) } }
end

COLUMNS = ENV["COLUMNS"] || 80

def singleton (&body)
  object = Object.new
  object.extend(Module.new(&body))
  object.send :initialize
  return object
end

status_line = singleton do
  attr_reader :width
  def initialize
    @width = 0
  end

  def remaining_width
    COLUMNS - @width
  end
  
  def clear
    print "\b" * COLUMNS
    print " " * COLUMNS
    print "\b" * COLUMNS
    @width = 0
  end

  def update
    clear ; yield ; flush
  end

  def flush
    $stdout.flush
  end
  
  def << string
    count = [remaining_width, string.size].min
    print string[0 ... count]
    @width += count
  end
end

n_total_files = 0
print "Counting files..." ; $stdout.flush
Find.find(*ARGV) { |x| n_total_files += 1 if FileTest.file? x }
puts " #{n_total_files}."

def warn_skip (file_name, message)
  puts "Warning: Skipping file `#{file_name}': #{message}"
end

n_completed_files = 0           # This counts all files.
n_processed_files = 0           # This only counts recognized files.
n_created_links = 0
Find.find *ARGV do |file_name|
  if FileTest.directory? file_name
    status_line.update do
      percent_done = n_completed_files * 100.0 / n_total_files
      status_line << "[%.2f%%] " % percent_done
      count = status_line.remaining_width - "Processing `'...".size
      if file_name.size > count
        file_name_tail = "[...]" + file_name[-count + 5 .. -1]
      else
        file_name_tail = file_name
      end
      status_line << "Processing `#{file_name_tail}'..."
    end
  elsif FileTest.file? file_name
    next if [".jpg", ".png", ".gif"].include? \
      File.extname(file_name).downcase
    begin 
      file = TagLib::File.new(file_name)
      n_processed_files += 1
      data = { :artist_name => file.artist,
               :album_year  => file.year.to_s,
               :album_title => file.album,
               :track_index => "#{0 if file.track < 10}#{file.track}",
               :track_title => file.title }

      for key, value in data do
        if value.blank?
          data.delete key
        else
          data[key] = value.trim
        end
      end

      components = parse_data(data).map { |x| join_components(*x) }.
        inject([]) { |a, x| a << join_components(a.last, x) }

      dir_name = components[0...-1] * "/"
      new_file_name = components * "/" + File.extname(file_name)

      FileUtils.mkdir_p(dir_name) if components.length > 1

      begin
        if $hardlinks
          FileUtils.ln(file_name, new_file_name)
        else
          FileUtils.ln_s(file_name, new_file_name)
        end
        n_created_links += 1
      rescue Errno::EEXIST
        if $hardlinks
          raise unless File.stat(file_name).ino ==
            File.stat(new_file_name).ino
        else
          raise unless FileTest.symlink? new_file_name and
            File.readlink(new_file_name) == file_name
        end
      end

    rescue TagLib::BadFile
      puts ; warn_skip file_name, "Unrecognized file format."
    rescue TagLib::BadTag
      puts ; warn_skip file_name, "Unreadable tag."
    rescue NotEnoughData
      puts ; warn_skip file_name, "Not enough track data " +
        "(need at least the track title)."
    rescue Errno::EEXIST
      puts ; warn_skip file_name,
        "Cannot create #{$hardlinks ? "hardlink" : "symbolic link"}: " +
        "Conflicting file already exists. [original error: #$!]"
    rescue Interrupt
      puts ; puts "Interrupted." ; exit(1)
    rescue Exception
      puts ; raise
    ensure
      file.close unless file == nil
    end

    n_completed_files += 1
  end
end

status_line.update do
  status_line << "[100%] Processing `#{ARGV.last}'..."
end

puts ; puts "Processed #{n_processed_files} media files " +
  "(created #{n_created_links} " +
  "#{$hardlinks ? "hardlinks" : "symbolic links"})."
